File: Islands/src/controlclim.jl
using NetCDF, MAT, Glob, Statistics, JLD2
using Seaborn

function ncname(fol::AbstractString,ii::Integer)
    fnc = "$(fol)/run000$(ii)/atmos_daily.nc"
end

function dataextractsfc(parisca::AbstractString,fol::AbstractString,nlon::Integer,nlat::Integer)

    cdir = pwd(); cd(fol); allfol = glob("run00*/"); cd(cdir); nfol = length(allfol);
    mdata = zeros(nlon,nlat,nfol-1); sdata = zeros(nlon,nlat,nfol-1);

    for ii = 2 : nfol
        mdata[:,:,ii-1] = mean(ncread(ncname(fol,ii),parisca),dims=3);
        sdata[:,:,ii-1] = std(ncread(ncname(fol,ii),parisca),dims=3);
    end

    mdata = mean(mdata,dims=3); mdata = mean(mdata,dims=1)[:]; mdata = (mdata + reverse(mdata))/2;
    sdata = mean(sdata,dims=3)/sqrt(nfol-1);
    sdata = mean(sdata,dims=1)[:]/sqrt(nlon); sdata = sdata + reverse(sdata);

    if parisca == "precipitation"; mdata = mdata*24*3600; sdata = sdata*24*3600;
    elseif parisca == "t_surf";    mdata = mdata .- 273.15;
    end

    return mdata,sdata

end

function dataextractlvl(parisca::AbstractString,fol::AbstractString,nlon::Integer,nlat::Integer)

    cdir = pwd(); cd(fol); allfol = glob("run00*/"); cd(cdir); nfol = length(allfol);
    mdata = zeros(nlon,nlat,30,nfol-1);

    for ii = 2 : nfol
        mdata[:,:,:,ii-1] = mean(ncread(ncname(fol,ii),parisca),dims=4);
    end

    mdata = mean(mdata,dims=4); mdata = mean(mdata,dims=1);
    mdata = (mdata+reverse(mdata,dims=2))/2; mdata = reshape(mdata,nlat,30);

    return mdata

end

reg = matread("lonlatpre.mat"); lon = reg["lon"][:]; lat = reg["lat"][:]; pre = reg["allpre"][:];
nlon = length(lon); nlat = length(lat);

RASfol = "/Volumes/CliNat-Isca/isca_out/Islands/T85L30-RRTM-RAS/control/";
QEBfol = "/Volumes/CliNat-Isca/isca_out/Islands/T85L30-RRTM-SBM/control/";

@load "controlprcp.jld2"
#mprcpRAS,sprcpRAS = dataextractsfc("precipitation",RASfol,nlon,nlat);
#mprcpQEB,sprcpQEB = dataextractsfc("precipitation",QEBfol,nlon,nlat);
#@save "controlprcp.jld2"  mprcpQEB mprcpRAS sprcpQEB sprcpRAS

close(); figure(figsize=(10,6),dpi=200)
plot(sind.(lat),mprcpRAS,label="Relaxed Arakawa-Schubert");
fill_between(sind.(lat),mprcpRAS-sprcpRAS,mprcpRAS+sprcpRAS,alpha=0.3);
plot(sind.(lat),mprcpQEB,label="Quasi-Equilibrium Betts-Miller");
fill_between(sind.(lat),mprcpQEB-sprcpQEB,mprcpQEB+sprcpQEB,alpha=0.3);
xlim(-1,1); ylim(0,10); grid("on")
xticks([-1,-sind(60),-sind(45),-sind(30),-sind(15),0,sind(15),sind(30),sind(45),sind(60),1],
       ["-90","-60","-45","-30","-15","0","15","30","45","60","90"]);
legend(); xlabel(L"Latitude / $\degree$"); ylabel(L"Precipitation / mm day$^{-1}$");
savefig("../figures/controlprcp.png",bbox_inches="tight")

@load "controltemp.jld2"
#mtempRAS,stempRAS = dataextractsfc("t_surf",RASfol,nlon,nlat);
#mtempQEB,stempQEB = dataextractsfc("t_surf",QEBfol,nlon,nlat);
#@save "controltemp.jld2"  mtempQEB mtempRAS stempQEB stempRAS

close(); figure(figsize=(10,6),dpi=200)
plot(sind.(lat),mtempRAS,label="Relaxed Arakawa-Schubert");
fill_between(sind.(lat),mtempRAS-stempRAS,mtempRAS+stempRAS,alpha=0.3);
plot(sind.(lat),mtempQEB,label="Quasi-Equilibrium Betts-Miller");
fill_between(sind.(lat),mtempQEB-stempQEB,mtempQEB+stempQEB,alpha=0.3);
xlim(-1,1); ylim(-50,30); grid("on")
xticks([-1,-sind(60),-sind(45),-sind(30),-sind(15),0,sind(15),sind(30),sind(45),sind(60),1],
       ["-90","-60","-45","-30","-15","0","15","30","45","60","90"]);
legend(); xlabel(L"Latitude / $\degree$"); ylabel(L"Temperature / $\degree$C");
savefig("../figures/controltemp.png",bbox_inches="tight")

@load "controluwind.jld2"
#uwindRAS = dataextractlvl("ucomp",RASfol,nlon,nlat);
#uwindQEB = dataextractlvl("ucomp",QEBfol,nlon,nlat);
#@save "controluwind.jld2" uwindQEB uwindRAS

close(); figure(figsize=(10,7),dpi=200)
contourf(sind.(lat),pre,uwindRAS');
xlim(-1,1); ylim(0,1000); grid("on")
xticks([-1,-sind(60),-sind(45),-sind(30),-sind(15),0,sind(15),sind(30),sind(45),sind(60),1],
       ["-90","-60","-45","-30","-15","0","15","30","45","60","90"]);
xlabel(L"Latitude / $\degree$"); ylabel(L"Pressure / hPa");
PyPlot.gca().invert_yaxis()
savefig("../figures/controluwind_RAS.png",bbox_inches="tight")

close(); figure(figsize=(10,7),dpi=200)
contourf(sind.(lat),pre,uwindQEB');
xlim(-1,1); ylim(0,1000); grid("on")
xticks([-1,-sind(60),-sind(45),-sind(30),-sind(15),0,sind(15),sind(30),sind(45),sind(60),1],
       ["-90","-60","-45","-30","-15","0","15","30","45","60","90"]);
xlabel(L"Latitude / $\degree$"); ylabel(L"Pressure / hPa");
PyPlot.gca().invert_yaxis()
savefig("../figures/controluwind_QEB.png",bbox_inches="tight")

File: Islands/src/controlclim.m
clear all; clc;

load lonlat.mat; nlon = numel(lon); nlat = numel(lat);
lon(end+1) = lon(1)+360; [ mlat,mlon ] = meshgrid(lat,lon);

RASfol = '/Volumes/CliNat-Isca/isca_out/Islands/T85L30-RRTM-RAS/';
QEBfol = '/Volumes/CliNat-Isca/isca_out/Islands/T85L30-RRTM-SBM/';

%%
prcpRAS = zeros(nlon,nlat,2); prcpQEB = zeros(nlon,nlat,2);
tempRAS = zeros(nlon,nlat,2); tempQEB = zeros(nlon,nlat,2);
for ii = 2 : 3
    
    prcpRAS(:,:,ii-1) = mean(ncread(ncname(RASfol,ii),'precipitation'),3);
    prcpQEB(:,:,ii-1) = mean(ncread(ncname(QEBfol,ii),'precipitation'),3);
    tempRAS(:,:,ii-1) = mean(ncread(ncname(RASfol,ii),'t_surf'),3);
    tempQEB(:,:,ii-1) = mean(ncread(ncname(QEBfol,ii),'t_surf'),3);
    
end

prcpRAS = mean(prcpRAS,3); prcpQEB = mean(prcpQEB,3);
tempRAS = mean(tempRAS,3); tempQEB = mean(tempQEB,3);

%%
prcpRAS = prcpRAS * 24*3600; prcpQEB = prcpQEB * 24*3600;
tempRAS = tempRAS - 273.15;  tempQEB = tempQEB - 273.15;

prcpRAS = hemsym(prcpRAS); prcpQEB = hemsym(prcpQEB);
tempRAS = hemsym(tempRAS); tempQEB = hemsym(tempQEB);

prcpRAS = lonextend(prcpRAS); prcpQEB = lonextend(prcpQEB);
tempRAS = lonextend(tempRAS); tempQEB = lonextend(tempQEB);

%%
close all; fig_resize(1.9,1);
[ pcol,~,~ ] = cpal_extract('drywet',20,0,0);
[ tcol,~,~ ] = cpal_extract('RdBu_r',20,0,0);

subplot(4,5,[1,2,6,7]);
m_proj('mollweide','clon',180);
m_contourf(mlon,mlat,prcpRAS,0:0.5:10,'linestyle','none');
m_grid('tickdir','out','linewi',1,...
    'xtick',60:60:300,'ytick',-60:30:60,...
    'xticklabels',[],'yticklabels',[],'XaxisLocation','middle');
colormap(gca,pcol); caxis([0 10]);
title(['Relaxed Arakawa-Schubert' newline]);
ylabel('Precipitation / mm day^{-1}','fontweight','bold','fontsize',11);

subplot(4,5,[3,4,8,9]);
m_proj('mollweide','clon',180);
m_contourf(mlon,mlat,prcpQEB,0:0.5:10,'linestyle','none');
m_grid('tickdir','out','linewi',1,...
    'xtick',60:60:300,'ytick',-60:30:60,...
    'xticklabels',[],'yticklabels',[],'XaxisLocation','middle');
colormap(gca,pcol); caxis([0 10]);
title(['Quasi-Equilibrium Betts-Miller' newline]);

subplot(4,5,[11,12,16,17]);
m_proj('mollweide','clon',180);
m_contourf(mlon,mlat,tempRAS,-50:5:50,'linestyle','none');
m_grid('tickdir','out','linewi',1,...
    'xtick',60:60:300,'ytick',-60:30:60,...
    'xticklabels',[],'yticklabels',[],'XaxisLocation','middle');
colormap(gca,tcol); caxis([-50 50]);
ylabel(['Surface Temperature / ' sprintf('%c',char(176)) 'C'],...
    'fontweight','bold','fontsize',11);

subplot(4,5,[13,14,18,19]);
m_proj('mollweide','clon',180);
m_contourf(mlon,mlat,tempQEB,-50:5:50,'linestyle','none');
m_grid('tickdir','out','linewi',1,...
    'xtick',60:60:300,'ytick',-60:30:60,...
    'xticklabels',[],'yticklabels',[],'XaxisLocation','middle');
colormap(gca,tcol); caxis([-50 50]);

subplot(4,5,[5,10]);
colormap(gca,pcol); caxis([0 10]); colorbar('west'); axis off;

subplot(4,5,[15,20]);
colormap(gca,tcol); caxis([-50 50]); colorbar('west'); axis off;

saveas(gcf,'controlclim','png');

%%

function fnc = ncname (fol,ii)
jj = num2str(ii); fnc = [ fol './control/run000' jj '/atmos_daily.nc' ];
end

function data = hemsym (data)
data = (data + fliplr(data))/2;
end

function data = lonextend(data)
data(end+1,:) = data(1,:);
end